Refactor ComplexMethod

arturbosch opened the pull request:
Now considers
- &&, || and ?: everywhere
- break, continue (throw is explicitly not counted like in ndepend)
- functions used to replace `if` or `for` are now configurable
- threshold increased to 15 like ndepend does - also linked their page

Closes #1921

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethod.kt:
```diff
@@ -46,47 +51,61 @@ class ComplexMethod(
         }
-        return false
+        // the case where function-expression syntax is used: `fun test() = when { ... }`
+        bodyExpression is KtWhenExpression -> true
```
Do we want to have a test for this case?

arturbosch replied:
Yep we already have some in the end of the Spec file.

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethod.kt:
```diff
@@ -46,47 +51,61 @@ class ComplexMethod(
     private val ignoreSimpleWhenEntries = valueOrDefault(IGNORE_SIMPLE_WHEN_ENTRIES, false)
+    private val ignoreNestingFunctions = valueOrDefault(IGNORE_NESTING_FUNCTIONS, false)
+    private val nestingFunctions =
+        CommaSeparatedPattern(valueOrDefault(NESTING_FUNCTIONS, ""), delimiters = ",;")
```
I think this is the first config option that can have a semicolon `;` as a delimiter.
Shouldn't we also stick to the standard delimiter for this case?

arturbosch replied:
You are right, I thought the `SplitPattern` does support `,;` by default. Will change this, thx!

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethod.kt:
```diff
@@ -21,22 +22,26 @@ import org.jetbrains.kotlin.psi.KtWhenExpression
  * @configuration ignoreSingleWhenExpression - Ignores a complex method if it only contains a single when expression.
  * (default: `false`)
  * @configuration ignoreSimpleWhenEntries - Whether to ignore simple (braceless) when entries. (default: `false`)
+ * @configuration ignoreNestingFunctions - Whether to ignore functions which are often used instead of an 'if' or
```
```suggestion
 * @configuration ignoreNestingFunctions - Whether to ignore functions which are often used instead of an `if` or
```

schalkms commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethod.kt:
```diff
@@ -21,22 +22,26 @@ import org.jetbrains.kotlin.psi.KtWhenExpression
  * (default: `false`)
  * @configuration ignoreSimpleWhenEntries - Whether to ignore simple (braceless) when entries. (default: `false`)
+ * @configuration ignoreNestingFunctions - Whether to ignore functions which are often used instead of an 'if' or
+ * 'for' statement (default: `false`)
```
```suggestion
 * `for` statement (default: `false`)
```

schalkms commented on detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/ComplexMethodSpec.kt:
```diff
@@ -11,20 +11,49 @@ import org.spekframework.spek2.Spek
             it("finds one complex method") {
-                val subject = ComplexMethod()
-                subject.lint(Case.ComplexMethod.path())
+                val findings = ComplexMethod(threshold = 10).lint(Case.ComplexMethod.path())
```
This case file is still a pain in the ass. `Case.ComplexMethod.path()`

schalkms replied:
However, this should be done in a separate PR.

schalkms commented on detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/internal/CyclomaticComplexity.kt:
```diff
@@ -0,0 +1,130 @@
+        }
+    }
+
+    private fun isInsideObjectLiteral(function: KtNamedFunction): Boolean =
```
```suggestion
    private fun isInsideObjectLiteral(function: KtNamedFunction) =
```

arturbosch commented:
> I particularly like the config implementation for the CyclomaticComplexity class.
> TIL that you can now use Kotlin functions inside Spec's `context()` functions. Cool!
> As I commented, the `ComplexMethod` case file is very hard to understand and maintain.
> Hopefully, we can also approach at some point in the future.
> For now, this PR looks very fine!

Thanks! I've refactored some test cases and removed that ComplexMethod case file :)
